packages/taiko-client/prover/proof_submitter/proof_buffer_monitor.go

```diff
@@ -49,3 +54,134 @@ func monitorProofBuffer(
 		}
 	}
 }
+
+// startCacheCleanUpAndFlush launches goroutines that keep cache maps pruned from
+// already-finalized proposal IDs and flush cache into buffers.
+func startCacheCleanUpAndFlush(
+	ctx context.Context,
+	rpc *rpc.Client,
+	proofCacheMaps map[proofProducer.ProofType]*ProofCache,
+	proofBuffers map[proofProducer.ProofType]*proofProducer.ProofBuffer,
+) {
+	log.Info("Starting proof cache cleanup and flushing monitors", "monitorInterval", monitorInterval)
+	for proofType, cacheMap := range proofCacheMaps {
+		buffer, ok := proofBuffers[proofType]
+		if !ok {
+			log.Error("Proof buffer for proof type not found", "proofType", proofType)
+			return
+		}
+		go cleanUpStaleCacheAndFlush(ctx, rpc, cacheMap, monitorInterval, buffer)
+	}
+}
+
+// cleanUpStaleCacheAndFlush periodically removes cached proofs that have already been
+// finalized so stale entries do not accumulate and flush cache into buffer.
+func cleanUpStaleCacheAndFlush(
+	ctx context.Context,
+	rpc *rpc.Client,
+	cacheMap *ProofCache,
+	cleanUpInterval time.Duration,
+	buffer *proofProducer.ProofBuffer,
+) {
+	ticker := time.NewTicker(cleanUpInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			log.Debug("context of proof cache cleanup monitor is done")
+			return
+		case <-ticker.C:
+			coreState, err := rpc.GetCoreStateShasta(&bind.CallOpts{Context: ctx})
+			if err != nil {
+				log.Error("Failed to get Shasta core state", "error", err)
+				continue // Skip this iteration, retry on next tick
+			}
+			lastFinalizedProposalID := coreState.LastFinalizedProposalId
+			// remove stale cache
+			removeFinalizedProofsFromCache(cacheMap, lastFinalizedProposalID)
+			// flush cached proofs
+			toID := lastFinalizedProposalID.Uint64() + buffer.MaxLength
+			if err := flushProofCacheRange(lastFinalizedProposalID.Uint64(), toID, buffer, cacheMap); err != nil {
+				if !errors.Is(err, ErrCacheNotFound) {
+					log.Error("Failed to flush proof cache range",
+						"error", err,
```

Comment:
```suggestion
					log.Error(
					    "Failed to flush proof cache range",
						"error", err,
```

Let's maintain a consistent logging style in Go client.